services/src/handlers/model_training.rs

```diff
@@ -0,0 +1,154 @@
+use futures::StreamExt;
+use geoengine_datatypes::primitives::MachineLearningQueryRectangle;
+
+#[cfg(feature = "xgboost")]
+use geoengine_datatypes::pro::MachineLearningFeature;
+use geoengine_datatypes::raster::GridOrEmpty;
+use geoengine_operators::call_on_generic_raster_processor;
+use geoengine_operators::engine::{
+    ExecutionContext, InitializedRasterOperator, QueryContext, QueryProcessor, RasterOperator,
+    TypedRasterQueryProcessor,
+};
+use num_traits::AsPrimitive;
+
+use crate::error::Error;
+use crate::error::Result;
+use crate::workflows::workflow::Workflow;
+
+const BATCH_SIZE: usize = 1_000;
+
+#[derive(Debug)]
+enum MachineLearningAccumKind {
+    XGBoost(Vec<f32>),
+}
+
+#[cfg(feature = "xgboost")]
+#[derive(Debug)]
+/// Used to gather the source data before further processing/shaping for xgboost
+struct MachineLearningAccum {
+    feature_name: String,
+    accum: MachineLearningAccumKind,
+}
+
+impl MachineLearningAccum {
+    fn new(feature_name: String) -> MachineLearningAccum {
+        MachineLearningAccum {
+            feature_name,
+            accum: MachineLearningAccumKind::XGBoost(Vec::new()),
+        }
+    }
+
+    fn update(&mut self, values: impl Iterator<Item = f32>) {
+        match self.accum {
+            MachineLearningAccumKind::XGBoost(ref mut accumulator) => {
```

Comment:
can you implement this as a a trait instead of an enum? s.t. you don't have to match on the Kind for each update, but rather once at the beginning when choosing which accumulator to use.